sound/soc/sof/intel/hda.c

```diff
@@ -402,6 +402,28 @@ static const struct sof_intel_dsp_desc
 	return chip_info;
 }
 
+irqreturn_t hda_dsp_interrupt_handler(int irq, void *context)
+{
+	struct snd_sof_dev *sdev = context;
+
+	if (check_ipc(sdev) || is_stream_irq(sdev))
+		return IRQ_WAKE_THREAD;
+
+	return IRQ_NONE;
+}
+
+irqreturn_t hda_dsp_interrupt_thread(int irq, void *context)
+{
+	struct snd_sof_dev *sdev = context;
+
+	if (check_ipc(sdev))
+		sof_ops(sdev)->irq_thread(irq, sdev);
+	if (is_stream_irq(sdev))
+		hda_dsp_stream_threaded_handler(irq, sdev);
+
+	return IRQ_HANDLED;
+}
```

Comment:
@bardliao this looks bad. You call check_ipc() in the handler, and then wake up the thread and then check_ipc() again and do the same thing over?

Reply:
@ranj063 Yes, that's odd, but there are more than one interrupt could be trigger and one may be triggered after the handler is checked. So we need to recheck it on the thread. Actually, I think we can have a flag that store the interrupt status which is checked on the handler. So we won't need to double check it on the thread again. Please see [hda_dsp_irq_thread](https://github.com/thesofproject/linux/pull/1451/files#diff-0c41ff728d2838606e1e70839d291aedR294)

Reply:
Why not just return IRQ_WAKE_THREAD in hda_dsp_interrupt_handler() and then handle all the checks in the thread?

Reply:
> Why not just return IRQ_WAKE_THREAD in hda_dsp_interrupt_handler() and then handle all the checks in the thread?

@ranj063 We can only do it when there is not a shared irq. i.e. No one else could trigger the irq. Probably, we can do that after ipc/stream/sdw irq are merged. And we don't need the handler if it return IRQ_WAKE_THREAD always.